### ISSUE TITLE
Indoor用データセットのダウンロードと学習環境構築

### ISSUE THREAD

**Issue by AtsushiHashimoto** [omron-sinicx/EfficientLoFTR#1]
## 概要

EfficientLoFTRのindoor用チェックポイントが公開されていないため（参照: [zju3dv/EfficientLoFTR#35](https://github.com/zju3dv/EfficientLoFTR/issues/35)）、自前でデータセットをダウンロードし、学習を行う。

## 背景

- オリジナルのEfficientLoFTRリポジトリでは、outdoor用のチェックポイントは提供されているが、indoor用は公開されていない
- 開発者によると、デモ動画ではScanNetデータセットで訓練されたカスタムチェックポイントを使用したが、公開予定はないとのこと
- したがって、indoor環境でEfficientLoFTRを使用するには、自前でScanNetデータセットを取得し学習する必要がある

### 1. データセット準備
- [ ] ScanNetデータセットのダウンロード手順確認
- [ ] ScanNet利用規約・アクセス申請
- [ ] データセットのダウンロード・前処理

### 2. 学習環境構築
- [ ] 学習スクリプトの確認・動作検証
- [ ] ハイパーパラメータの確認
- [ ] GPU環境の準備

### 3. 学習実行
- [ ] 学習の実行
- [ ] チェックポイントの検証

## 参考資料

- [EfficientLoFTR Issue #35](https://github.com/zju3dv/EfficientLoFTR/issues/35)
- [ScanNet Dataset](http://www.scan-net.org/)

**Comment by AtsushiHashimoto** [comment]
## タスク開始

- **ブランチ**: `research/1-indoor-training`
- **Worktree**: `worktrees/issue1`

作業を開始します。まずEfficientLoFTRの学習手順とScanNetデータセットの取得方法を調査します。

**Comment by AtsushiHashimoto** [comment]
## 進捗報告 (2026-01-07 08:19)

### 完了した作業

#### 1. pytorch-lightning 2.x への移行
公式リポジトリは pytorch-lightning 1.x 向けに書かれていたため、2.x への移行作業を実施:

| ファイル | 修正内容 |
|---------|---------|
| `train.py` | DDPPlugin→DDPStrategy、Trainer引数の手動設定化 |
| `src/lightning/lightning_loftr.py` | epoch_end→on_*_epoch_end、outputs管理方式変更 |
| `src/lightning/data.py` | 分散初期化チェック追加、単一GPU対応 |
| `src/utils/profiler.py` | import パス修正 |
| `src/utils/plotting.py` | `.detach()` 追加 |

#### 2. ScanNet 訓練用設定ファイル作成
- `configs/data/scannet_trainval.py` を新規作成

#### 3. データパスのシンボリックリンク作成
- `data/scannet/` → `/data/LoFTR/scannet*`
- `data/megadepth/` → `/data/LoFTR/megadepth*`

#### 4. 学習動作検証
- 1513シーンのロード成功
- 1エポックのテスト学習完了
- チェックポイント保存確認

### 変更ファイル
```
src/lightning/data.py            | 22 +++++++-----
src/lightning/lightning_loftr.py | 72 +++++++++++++++++++++++++++-------------
src/utils/plotting.py            |  6 ++--
src/utils/profiler.py            |  4 +--
train.py                         | 51 ++++++++++++++++++++--------
configs/data/scannet_trainval.py | (新規)
```

### 次のステップ
- [ ] ScanNet での本番学習を開始（nohup で実行）
- [ ] MegaDepth学習済みモデルとの比較レポート作成

**Comment by AtsushiHashimoto** [comment]
## 進捗報告 (2026-01-08 01:52) - Container Rebuild 前

### 完了した作業

#### 1. pytorch-lightning 2.x 移行 ✅
以下のファイルを修正して、pytorch-lightning 2.x に対応:
- `train.py`: DDPStrategy, Trainer引数の手動設定
- `src/lightning/lightning_loftr.py`: epoch_end → on_*_epoch_end
- `src/lightning/data.py`: 分散初期化チェック、単一GPU対応
- `src/utils/profiler.py`: import パス修正
- `src/utils/plotting.py`: .detach() 追加

#### 2. ScanNet 訓練設定 ✅
- `configs/data/scannet_trainval.py` 作成
- データシンボリックリンク設定済み

#### 3. 学習実行 (5エポック完了)
| Epoch | AUC@5° | AUC@10° | AUC@20° |
|-------|--------|---------|---------|
| 0 | 7.2% | 16.9% | 29.8% |
| 1 | 8.5% | 19.1% | 32.5% |
| **2** | **14.2%** | **29.3%** | **45.4%** ← ベスト |
| 3 | 7.6% | 17.4% | 29.7% |
| 4 | 7.6% | 17.3% | 29.1% |

**チェックポイント保存済み:**
```
logs/tb_logs/scannet_indoor_gpu0/version_0/checkpoints/last.ckpt
```

---

### 🔴 Container Rebuild 後にやるべきこと

#### Step 1: Worktree に移動
```bash
cd /workspaces/EfficientLoFTR/worktrees/issue1
```

#### Step 2: 4GPU で学習再開
```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 nohup python -u train.py \
    configs/data/scannet_trainval.py \
    configs/loftr/eloftr_full.py \
    --exp_name scannet_indoor_4gpu \
    --max_epochs 30 \
    --batch_size 16 \
    --num_workers 8 \
    --devices 4 \
    --ckpt_path logs/tb_logs/scannet_indoor_gpu0/version_0/checkpoints/last.ckpt \
    > logs/training_logs/scannet_indoor_4gpu_resumed.log 2>&1 &
```

#### Step 3: 進捗確認
```bash
# GPU使用状況
nvidia-smi

# ログ確認
tail -f logs/training_logs/scannet_indoor_4gpu_resumed.log

# チェックポイント確認
ls -la logs/tb_logs/scannet_indoor_4gpu/*/checkpoints/
```

---

### 予想される結果

| 訓練データ | AUC@5° | AUC@10° | AUC@20° |
|-----------|--------|---------|---------|
| MegaDepth (論文) | ~24% | ~43% | ~60% |
| **ScanNet (今回)** | **TBD** | **TBD** | **TBD** |

Indoor domain matching により +5-15% の改善を期待。

---

### 残タスク
- [ ] 4GPU で学習再開 (残り ~25 エポック)
- [ ] 学習完了後、MegaDepth学習モデルとの比較レポート作成
- [ ] チェックポイントの保存・共有

**Comment by AtsushiHashimoto** [comment]
## 進捗報告 (2026-01-08 02:03) - Container Rebuild 後の再開

### 実施した作業

#### 1. 環境復旧確認 ✅
- nvidia-smi 正常動作確認
- GPU 0-3 (A100 80GB x 4) 利用可能
- データセットシンボリックリンク確認済み

#### 2. チェックポイント確認 ✅
既存のチェックポイントが保存されていることを確認:
```
epoch=1-auc@5=0.085-auc@10=0.191-auc@20=0.325.ckpt
epoch=2-auc@5=0.142-auc@10=0.293-auc@20=0.454.ckpt (ベスト)
epoch=3-auc@5=0.076-auc@10=0.174-auc@20=0.297.ckpt
epoch=4-auc@5=0.076-auc@10=0.173-auc@20=0.291.ckpt
epoch=5-auc@5=0.124-auc@10=0.263-auc@20=0.414.ckpt
last.ckpt (epoch=5)
```

#### 3. 4GPU での学習再開 ✅
- `train.py` に `--resume_from_checkpoint` 引数を追加
- 以下のコマンドで学習を再開:

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 nohup python -u train.py \
    configs/data/scannet_trainval.py \
    configs/loftr/eloftr_full.py \
    --exp_name scannet_indoor_4gpu \
    --max_epochs 30 \
    --batch_size 4 \
    --num_workers 4 \
    --devices 4 \
    --resume_from_checkpoint logs/tb_logs/scannet_indoor_gpu0/version_0/checkpoints/last.ckpt \
    > logs/training_logs/scannet_indoor_4gpu_resumed.log 2>&1 &
```

### 現在の状態

| 項目 | 状態 |
|------|------|
| GPU使用 | 0-3 (各 ~11.5GB / 81920MB) |
| GPU稼働率 | 100% |
| 再開エポック | 6 (epoch 5 から継続) |
| 残りエポック | 6-29 (24 エポック) |

### 監視コマンド

```bash
# GPU状態
nvidia-smi

# ログ確認
tail -f logs/training_logs/scannet_indoor_4gpu_resumed.log

# チェックポイント
ls -la logs/tb_logs/scannet_indoor_4gpu/*/checkpoints/
```

### 次のステップ
- [ ] 学習完了まで監視
- [ ] 最終結果の報告
- [ ] MegaDepth学習モデルとの比較
